UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c

```diff
@@ -616,19 +617,49 @@ InitPaging (
   DEBUG ((DEBUG_INFO, "Patch page table start ...\n"));
   if (FeaturePcdGet (PcdCpuSmmProfileEnable)) {
     for (Index = 0; Index < mProtectionMemRangeCount; Index++) {
-      MemoryAttrMask = 0;
-      if (mProtectionMemRange[Index].Nx == TRUE) {
-        MemoryAttrMask |= EFI_MEMORY_XP;
-      }
+      Base   = mProtectionMemRange[Index].Range.Base;
+      Length = mProtectionMemRange[Index].Range.Top - Base;
 
+      MemoryAttrMask = EFI_MEMORY_RP;
       if (mProtectionMemRange[Index].Present == FALSE) {
-        MemoryAttrMask = EFI_MEMORY_RP;
+        //
+        // Set the EFI_MEMORY_RP attribute
+        //
+        IsSet = TRUE;
+      } else {
+        //
+        // Clear the EFI_MEMORY_RP attribute
+        //
+        IsSet = FALSE;
+
+        //
+        // Note: When mapping a range to present, EFI_MEMORY_RO and EFI_MEMORY_XP must be specified.
+        //
+        MemoryAttrMask |= EFI_MEMORY_RO;
+        MemoryAttrMask |= EFI_MEMORY_XP;
       }
 
-      Base   = mProtectionMemRange[Index].Range.Base;
-      Length = mProtectionMemRange[Index].Range.Top - Base;
-      if (MemoryAttrMask != 0) {
-        Status = ConvertMemoryPageAttributes (PageTable, mPagingMode, Base, Length, MemoryAttrMask, TRUE, NULL);
+      Status = ConvertMemoryPageAttributes (PageTable, mPagingMode, Base, Length, MemoryAttrMask, IsSet, NULL);
+      ASSERT_RETURN_ERROR (Status);
+
+      if (mProtectionMemRange[Index].Present == TRUE) {

```

Comment:
If (Present && Nx) {}
Because above logic already clears NX bit, "Present && !Nx" is not needed to handle here.